freqtrade/freqai/data_drawer.py

```diff
@@ -155,14 +165,22 @@ def save_historic_predictions_to_disk(self):
         # create a backup
         shutil.copy(self.historic_predictions_path, self.historic_predictions_bkp_path)
 
-    def save_drawer_to_disk(self):
+    def save_drawer_to_disk(self, live=False):
         """
         Save data drawer full of all pair model metadata in present model folder.
         """
-        with self.save_lock:
+        if live:
+            with self.save_lock:
+                with open(self.pair_dictionary_path, 'w') as fp:
+                    rapidjson.dump(
+                                    self.pair_dict, fp, default=self.np_encoder,
+                                    number_mode=rapidjson.NM_NATIVE)
+        else:
+            # save_lock it's not working with hyperopt
             with open(self.pair_dictionary_path, 'w') as fp:
-                rapidjson.dump(self.pair_dict, fp, default=self.np_encoder,
-                               number_mode=rapidjson.NM_NATIVE)
+                rapidjson.dump(
```

Comment:
i don't think this will really work / suffice (at least not without introducing a TON more problems).

this will only work if the freqAI model is guaranteed to not retrain the models - and in the end, not to not call this method.
Hyperopt runs backtesting in several parallel procesess.
This means, you now potentially have X parallel freqAI trainings running (if using `--analyze-per-epoch`, or implementing `freqai.start()` in a  non-default place - see below)
apart from the fact that these will slow each other down as they each scale out to multiple threads/processes themselfs - they'll also write the model to disk once done.
They'll however write the model to disk - using the **same** filename - over and over again.

Ideally - you'll "just" have models that change all the time - but there's nothing preventing the model for ETH/USDT from being written to disk at the same time - causing corruption and/or crashes.

---

now in theory we can just reject to start hyperopt on a freqAI strategy that also defines `--analyze-per-epoch` - but i think a "2nd layer of defense" should be added - to protect from users implementing freqai not in `populate_indicators()` - but in `populate_entry_trend()` - which is called in parallel processes all the time, independent of `--analyze-per-epoch`.

Reply:
After a better analysis, we don't need to call this method in hyperopt (in parallels process). Only in the first run, if the user choose to save models, but this happens before the parallels execution. All things inside `populate_any_indicators` in a Freqai strategy will no be hyperopt. The root problem is that I was using the FreqaiExampleStrategy and we have this call inside the strategy, in `confirm_trade_exit` and `custom_exit` methods:

```
        pair_dict[pair]['extras']["prediction" + entry_tag] = 0
        if not follow_mode:
            self.freqai.dd.save_drawer_to_disk()
        else:
            self.freqai.dd.save_follower_dict_to_disk()
```

@robcaulk , why do we need this calls inside the strategy? This is handled by the backend, right? Maybe something old that we need to clear.

Reply:
The only goal of this code is to use the trade candle prediction for exit logic. We store it to disk because we want crash resilience. 

This means that it can be removed for backtesting, and the same info can be stored using a dictionary similar to `custom_info` example in the docs: https://www.freqtrade.io/en/stable/strategy-advanced/#storing-information

Or just remove it entirely and let users handle this on their own.

Reply:
Thanks Rob. Great. My suggestion for now is to remove this in example strategy and we can keep a task to add a new example using `custom_info`. I will continue my tests without this code.

Reply:
I never liked this code being in the example strategy. We can remove it from the `custom_exit` and the `confirm_trade_exit` . The example strategy should be simple and this just muddies the waters.

Reply:
I will remove all code in example strategy that refers to self.freqai.dd

Reply:
> All things inside populate_any_indicators in a Freqai strategy will no be hyperopt.

that's ... not really true - it's an assumption based on the default/sample strategy - but nothing prevents me (or any other user) from adding another indicator that's based on a hyperoptable parameter to populate_any_indicators, instead of using what's currently defined.

While we do loop on `indicator_periods_candles` (for example) - that's a parameter that could for sure be hyperopted (which would however require retraining the model each time - maybe even requiring a new identifier ...?)
It's not used anywhere internally other than for candle startup calcs - the rest must be done within the strategy - so no-one forces me to use this - instead of a hyperoptable parameter (or set of parameters) to experiment with different lengths.

it's just a sample, really - the  real point is - nothing prevents the usage of parameters in `populate_any_indicators` ...

Reply:
we sure can - my point is - we have the samples that call it in one place - but nothing says "we don't support other locations".
Documentation can cover a lot of things - maybe we can add a sample in the freqAI docs about the limits of hyperopt specifically? so it's grouped together ...

Reply:
> I got it! I know that a user can in another strategy try to hyperopt an indicator in `populate_any_indicators`, but I think that we need to inform in docs that this will not work (at this moment). If we hyperopt in `populate_any_indicators`, the process will be extremely slow, because it will require a new training for each combination. I think that the main goal, at this moment, is to allow user hyperopt all other options/spaces except `populate_any_indicators`. This will be very helpful, even if it's not a complete solution yet.
> 
> > now in theory we can just reject to start hyperopt on a freqAI strategy that also defines --analyze-per-epoch -
> 
> I agree. This will be necessary.
> 
> > to protect from users implementing freqai not in populate_indicators() - but in populate_entry_trend()
> 
> What do you mean? Prevent an user to call `self.freqai.start()` in populate_entry_trend()? Couldn't we handle this in the documentation?

Yes, we cannot allow hyperopt to retrain models. It adds complexity and it is anyways not computationally tractable.

If they want to hyperopt with FreqAI, they are limited to entry/exit criteria. That should simplify this PR and will also reduce the number of questions we can expect in discord about "why the hyperopt is taking so long?"

Reply:
> we sure can - my point is - we have the samples that call it in one place - but nothing says "we don't support other locations".
@robcaulk , do you see any situation that will be necessary to call self.freqai outside `populate_indicators()` or `populate_any_indicators()`

Reply:
No. 

Calling `self.freqai.start()` inside `populate_any_indicators()` would cause an infinite loop. 

We haven't seen users trying to call `self.freqai.start()` anywhere besides the instructed location of `populate_indicators()`. So I don't think this is much of a concern at the moment.

Reply:
Well. New version with all considerations addressed!

Reply:
In the end, the change was much simpler than I expected.